src/fsharp/service/IncrementalBuild.fs

```diff
@@ -1022,16 +1023,91 @@ module IncrementalBuilderEventTesting =
 
 module Tc = FSharp.Compiler.TypeChecker
 
+[<AutoOpen>]
+module IncrementalBuildSyntaxTree =
 
-/// Accumulated results of type checking.
+    [<Sealed>]
+    type SyntaxTree (tcConfig: TcConfig, fileParsed: Event<string>, lexResourceManager, sourceRange: range, filename: string, isLastCompiland) =
+
+        let mutable weakCache: WeakReference<_> option = None
+
+        let parse(sigNameOpt: SyntaxTree.QualifiedNameOfFile option) =
+            let errorLogger = CompilationErrorLogger("Parse", tcConfig.errorSeverityOptions)
+            // Return the disposable object that cleans up
+            use _holder = new CompilationGlobalsScope(errorLogger, BuildPhase.Parse)
+
+            try  
+                IncrementalBuilderEventTesting.MRU.Add(IncrementalBuilderEventTesting.IBEParsed filename)
+                let lower = String.lowercase filename
+                let canSkip = sigNameOpt.IsSome && FSharpImplFileSuffixes |> List.exists (Filename.checkSuffix lower)
+                let input = 
+                    if canSkip then
+                        SyntaxTree.ParsedInput.ImplFile(
+                            SyntaxTree.ParsedImplFileInput(
+                                filename, 
+                                false, 
+                                sigNameOpt.Value,
+                                [],
+                                [],
+                                [],
+                                isLastCompiland
+                            )
+                        ) |> Some
+                    else
+                        ParseOneInputFile(tcConfig, lexResourceManager, [], filename, isLastCompiland, errorLogger, (*retryLocked*)true)
+
+                fileParsed.Trigger filename
+
+                let res = input, sourceRange, filename, errorLogger.GetErrors ()
+                // If we do not skip parsing the file, then we can cache the real result.
+                if not canSkip then
+                    weakCache <- Some(WeakReference<_>(res))
+                res
+            with exn -> 
+                let msg = sprintf "unexpected failure in SyntaxTree.parse\nerror = %s" (exn.ToString())
+                System.Diagnostics.Debug.Assert(false, msg)
+                failwith msg
+
+        /// Parse the given file and return the given input.
+        member _.Parse sigNameOpt =
+            match weakCache with
+            | Some weakCache ->
+                match weakCache.TryGetTarget() with
+                | true, res -> res
+                | _ -> parse sigNameOpt
+            | _ -> parse sigNameOpt
+
+        member _.FileName = filename
+
+/// Accumulated results of type checking. The minimum amount of state in order to continue type-checking following files.
 [<NoEquality; NoComparison>]
-type TypeCheckAccumulator =
-    { tcState: TcState
-      tcImports: TcImports
-      tcGlobals: TcGlobals
-      tcConfig: TcConfig
-      tcEnvAtEndOfFile: TcEnv
+type TcInfo =
```

Comment:
`TcInfo` and `TcInfoOptional` is a placeholder name. I'm not quick sure what to name them yet.

Reply:
Actually, I think the names are as good as they are going to get. Could set it back to `TcAcc` but I prefer `Info`. This is all implementation detail.